op-supervisor/supervisor/backend/db/logs/db.go

```diff
@@ -39,9 +30,168 @@ type Metrics interface {
 	RecordDBSearchEntriesRead(count int64)
 }
 
+// logContext is a buffer on top of the DB,
+// where blocks and logs can be applied to.
 type logContext struct {
-	blockNum uint64
-	logIdx   uint32
+	// current entry index, including the contents of `out`
+	entryIndex entrydb.EntryIdx
+
+	// current block data
+	blockHash types.TruncatedHash
+	blockNum  uint64
+	timestamp uint64
+	doneBlock bool // when the block data has been fully appended to out
+
+	// current log event data
+	logIdx  uint32
+	logHash types.TruncatedHash
+	execMsg *types.ExecutingMessage
+	doneLog bool // when the log data has been fully appended to out
+
+	// last type of entry we added
+	lastType entrydb.EntryType
+
+	// buffer of entries not yet in the DB
+	out []entrydb.Entry
+}
+
+type EntryObj interface {
+	encode() entrydb.Entry
+}
+
+func (l *logContext) BlockHash() types.TruncatedHash {
+	if !l.doneBlock {
+		panic("logContext in invalid state, cannot retrieve block ID")
+	}
+	return l.blockHash
+}
+
+// TODO other getters, with checks if doneBlock and doneLog
+
+func (l *logContext) applyEntry(entry entrydb.Entry) {
+	if len(l.out) != 0 {
+		panic("can only apply without appending if the state is still empty")
+	}
+
+	// TODO decode entry
+
+	// TODO apply to state
+
+	l.lastType = entry.Type()
+	l.entryIndex += 1
+}
+
+func (l *logContext) appendEntry(obj EntryObj) {
+	entry := obj.encode()
+	l.out = append(l.out, entry)
+	l.lastType = entry.Type()
+	l.entryIndex += 1
+}
+
+func (l *logContext) infer() error {
```

Comment:
```suggestion
// infer advances the logContext in cases where multiple entries are to be appended implicitly
// depending on the last type of entry, a new entry is appended, or when the searchCheckpoint should be inserted
// This can be done repeatedly until there is no more implied data to extend 
func (l *logContext) infer() error {
```